### PR TITLE
Adjusting code interface

### PR DESCRIPTION
**Pull request by Roy-AMD**
glslang/include/intermediate.h -> Add a new interface to set TIntermBranch's expression.

glslang/include/Types.h -> Add interface to set Type's basicType and add interface to get basicType form a TSampler.

glslang/MachineIndependent/intermediate.cpp -> Part of the code in createConversion been encapsulating as a new function called buildConvertOp

glslang/MachineIndependent/localintermediate.h -> Export createConversion and
buildConvertOp as a public function, and add a new param to control whether force update xbfBufferStride.

glslang/Public/ShaderLang.h -> Add interface to get shader object and shader source.

This change are use for shader dump and Patch AST, we want do some patch on AST when after parser and before SPVGen, add more interface and export some interface to public, can make code scalability better.

### CODE REVIEW

File: glslang/Include/Types.h [hunk 1]
```diff
@@ -2184,7 +2186,8 @@ class TType {
     const TTypeList* getStruct() const { assert(isStruct()); return structure; }
     void setStruct(TTypeList* s) { assert(isStruct()); structure = s; }
     TTypeList* getWritableStruct() const { assert(isStruct()); return structure; }  // This should only be used when known to not be sharing with other threads
-
+    void setBasicType(const TBasicType& t) { basicType = t; }
```
Comment (johnkslang) [inline glslang/Include/Types.h]: I also don't see any calls to this, and it seems potentially a bit dangerous to expose this like this.

Is it getting used somewhere?

(The one in TSampler should probably also be protected; that struct has evolved a lot to the point it should probably be a class.  But, first would like to understand more what's going on here.)

File: glslang/MachineIndependent/Intermediate.cpp [hunk 2]
```diff
@@ -827,11 +783,64 @@ TIntermTyped* TIntermediate::createConversion(TBasicType convertTo, TIntermTyped
...
+    }
+
+    if (! getArithemeticFloat16Enabled()) {
+        if ((convertTo == EbtFloat16 && ! convertFromFloatTypes) ||
+            (node->getBasicType() == EbtFloat16 && ! convertToFloatTypes))
+            return nullptr;
+    }
+
+    if (!buildConvertOp(convertTo, node->getBasicType(), newOp))
+    {
```
Comment (johnkslang) [inline glslang/MachineIndependent/Intermediate.cpp]: Inconsistent curly braces.

File: glslang/MachineIndependent/localintermediate.h [hunk 3]
```diff
@@ -693,9 +693,9 @@ class TIntermediate {
     bool getPixelCenterInteger() const { return pixelCenterInteger; }
     void addBlendEquation(TBlendEquationShift b) { blendEquations |= (1 << b); }
     unsigned int getBlendEquations() const { return blendEquations; }
-    bool setXfbBufferStride(int buffer, unsigned stride)
+    bool setXfbBufferStride(int buffer, unsigned stride, bool force = false)
```
Comment (johnkslang) [inline glslang/MachineIndependent/localintermediate.h]: It almost seems like you want a changeXfbBufferStride().  Perhaps a changeBasicType() in TType as well?

Why would these things be changing in the middle of something?  It seems contrary to proper semantic checking, and fragile WRT all class invariants that are dependent of the member's value.

File: glslang/Include/Types.h (hunk 1 above)
Comment (Roy-AMD) [reply glslang/Include/Types.h]: It's for some external code. we try to patch on AST and modify it's basic type, such as modify a lowp float to float16_t,  or a sampler to f16sampler, I think it's not a general feature so I didn't upstream this code. open this interface can help us easy to do local patch on AST.

File: glslang/MachineIndependent/localintermediate.h (hunk 3 above)
Comment (Roy-AMD) [reply glslang/MachineIndependent/localintermediate.h]: The parameter are thoughtless, we will try to found a better way to fix it in local, I will revert it.

File: glslang/MachineIndependent/Intermediate.cpp (hunk 2 above)
Comment (Roy-AMD) [reply glslang/MachineIndependent/Intermediate.cpp]: fixed

**Comment by Roy-AMD** [comment]
Hi @johnkslang , code is updated, please take a look.

**Comment by Roy-AMD** [comment]
Hi @johnkslang ,

If there are no more questions, please merge this code.

Thanks
Roy

**Pull request merged**